Initial virtual folders

unho opened the pull request:
Ready to review.

This is meant to somehow recover our old goals functionality, but using the new vfolders.

This provides a management command to create and update virtual folders, adds the ability to sort the units by (vfolder) priority on the editor, and injects the priority sorting into the translate URLs in the translation project overview.

Things that differ from specs:

- We don't use any filtering rules. Instead we provide lists of filenames relative to vfolder location.

dwaynebailey commented:
@unho I thought we were going to change to use floats for the priority to allow 0.N<1(default)<N

To reviewers note:
There is no vfolder UI.  The translators are directed by priority if they choose to translate, address checks or reviews.  UI will come later but for now we can direct translators to work on the top priority vfolders.

@julen @iafan this needs your input I think especially since we now pass ordering info in the URLs.

unho commented:
@dwaynebailey If I remember correctly we discussed which changes would be necessary in case we actually decided to do such change. We didn't agree on doing such change, and I strongly disagree.

julen commented:
Any pointers for testing?

Can we have some tests as well, please?

julen commented on pootle/templates/editor/_toolbar.html:
```diff
@@ -4,6 +4,7 @@
   <div class="unit-sort">
     <select id="js-filter-sort">
       <option selected value="default">{% trans "Default order" %}</option>
+      <option value="priority">{% trans "Decreasing priority" %}</option>
```
Does this mean *higher priority*?

unho replied:
It means top priority on the top which decreases as we approach to the bottom.

julen commented on pootle/core/browser.py:
```diff
@@ -100,7 +100,12 @@ def make_generic_item(path_obj):
-    item = make_generic_item(directory)
+    filters = {}
+
+    if len(VirtualFolder.get_matching_for(directory.pootle_path)):
```
As I see this will incur in N*M+1 queries, where:
* N is the number of directories being displayed in the overview,
* M is the number of filtering rules in each browsable `VirtualFolder`, and
* 1 is the query needed for retrieving browsable `VirtualFolder`s

Looks excessive on a first sight, even more taking into account there's no caching of any sort.

unho replied:
This has been replaced by a new call that triggers only 1 query.

Regarding the cache, the plan is to get something that works, and then optimize it using cache.

dwaynebailey commented:
@unho I assume you disagree with float for priority?  Right, so this came up in the F2F.  We discussed why it was needed because of 0.N level priorities.  You said we can't do 0.N because we're using ints.  @jleclanche said we could simply use floats to achieve that.  So what are we doing to achieve 0.N type priorities? Apart from that you said nothing about disagreeing, especially since I'd articulated why it was needed. You really do need to make yourself heard, especially if we're doing a F2F.

unho commented:
@dwaynebailey I strongly disagree with using floats. According to https://docs.djangoproject.com/en/1.7/ref/models/fields/#positiveintegerfield we have a range of positive integers between 0 and 2147483647 to specify the priority. I see no need to use floats, that according to your comments on the specs doc can confuse people that are not good at maths. Reading the specs, the use of floats comes from some use of priority for scoring, but IMHO that can be achieved by multiplying the priority by some multiplier like 0.1 (yes, a float).

dwaynebailey commented:
@unho how are we addressing goals that have a priority of below normal?

In most projects you would have:

1. Some high priority goals
2. Units/Stores that have no special priority set at all.
3. Some items you want to be done last

unho commented:
@dwaynebailey I have been working on altering this to allow priorities between 0 and 1 as told on the specs https://docs.google.com/document/d/16g_ustDX2Pxm8yasqIw_kXXi2g1oU25iOnjOBCYcYWc

@julen I have looking at adding tests, but running py.test with InnoDB is proving not to be so easy. So since this looks like it is going to be time consuming, I am going to leave tests for this feature out for now.

julen commented:
@unho if you are running into trouble when running tests I think it's wiser to come and ask for help instead of disregarding tests altogether.

unho commented:
Cleaned the PR. Ready to review again.

It still has no tests. I might look into that tomorrow.

I am working on a list of steps to locally verify that this works. Maybe tomorrow.

unho commented:
In order to check this locally you will have to follow these steps:

- Check out the branch locally
- Ensure you have a project with several files on it (per language). It is recommended to have enough files to include some of them in a vfolder while leaving some of them out
- Run migrations
- Craft a JSON file to add some sample vfolders. Example in https://github.com/translate/pootle/pull/3600/files#diff-32d9c83230475a8ff37b1695b1606581R386 It is recommended to have a directory with 4 files at least, and include 3 of them on vfolders (one vfolder with 3 files, and the other two vfolders with 2 files, sharing only one of them).
- Run `add_vfolders` management command to import such vfolders. Example in https://github.com/translate/pootle/pull/3600/files#diff-32d9c83230475a8ff37b1695b1606581R456
- Run the server and go to a path where we know there is some of the files matched by a vfolder
- Check that the translate links above the stats table have the `sort=priority` bit on the URLs
- Check that the translate links on the stats table for the files or directories with files included in a vfolder also include that `sort=priority` bit on the URLs
- Click on any of those translate links and ensure the units are sorted by priority (since we are currently including the units on vfolders using the stores, we just need to ensure that the vfolders are in the appropriate order).

unho commented:
@dwaynebailey @iafan Please review the PR (or rereview it).

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+.. versionadded:: 2.7
+
+Virtual folders provide a way to group translations based on any criteria,
+including a same file across all the languages in a project, or specific files.
```
What does *a same file* mean here? Can you maybe reword it?

unho replied:
It means that a file named `file.po` is matched in all languages in a project. For example `/af/proj/file.po` or `/eu/proj/file.po`.

Rephrased to avoid confusion.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+- An optional description,
+- An optional filtering rule.
+
+Each virtual folder must have a different name and location pair. This means
```
Do you mean that name + location must be unique? or that name and location must differ?

unho replied:
The former one. Rewritten.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+- An optional filtering rule.
+
+Each virtual folder must have a different name and location pair. This means
+that there can exist two different virtual folders with the same if they have
```
*name* is missing after *with the same*.

unho replied:
Fixed. Thanks!

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+The priority defaults to ``1`` and accepts any value greater than ``0``,
+including numbers with decimals, like ``0.75``.
+
+By default the virtual folders are not browsable. This means that are not
```
What about: *By default virtual folders cannot be browsed: they won't be displayed and it won't be possible to use them for sorting.*

unho replied:
It is better. Picked it. Thanks.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+
+Also the virtual folders can have a description which might be useful to tell
+the translators which is the grouping criteria or provide additional
+instructions. This might be handy when using the virtual folders as goals.
```
I don't believe you need to use the article as in *the virtual folders*. @dwaynebailey?

unho replied:
Maybe I am wrong, but I think it is ok.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+the translators which is the grouping criteria or provide additional
+instructions. This might be handy when using the virtual folders as goals.
+
+The filtering rule tells which translation units are included within a virtual
```
What about: *The filtering rule specifies*

unho replied:
Better. Picked. Thanks.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+
+The filtering rule tells which translation units are included within a virtual
+folder. It is meant to be used internally in order to calculate the units that
+are included within the virtual folder. Currently it consists on a list of file
```
*consists of*

unho replied:
Think in galician, write in english, read @julen comments. :)

Fixed. Thanks!

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+instructions. This might be handy when using the virtual folders as goals.
+
+The filtering rule tells which translation units are included within a virtual
+folder. It is meant to be used internally in order to calculate the units that
```
This sentence doesn't seem to add any extra information?

unho replied:
Removed. Thanks!

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,69 @@
+If any virtual folder applies in the current location, the links on the
+overview will automatically point to the translation editor providing the
+priority sorting flag. The priority sorting on the translation editor is
+calculated having in account all the applicable virtual folders in the current
```
*taking into account*

unho replied:
Fixed. Thanks.

julen commented on docs/server/commands.rst:
```diff
@@ -346,6 +346,129 @@ By default, :ref:`commands#test_checks` tests all existing checks. When
+specified virtual folders already exist and the data differs, then they are
+updated with the provided data.
+
+The JSON file consists in a list of virtual folder definitions with the
```
*consists of*

unho replied:
Fixed. Thanks!

julen commented on docs/server/commands.rst:
```diff
@@ -346,6 +346,129 @@ By default, :ref:`commands#test_checks` tests all existing checks. When
+updated with the provided data.
+
+The JSON file consists in a list of virtual folder definitions with the
+following fields:
```
Can we split the expected format docs away from the command reference and put it in its own section? Since there's already a *docs/features/virtual_folders.rst* file and nothing links to the format spec, it's rather surprising to find this information along the commands.

unho replied:
@julen Since this JSON file is tied to the command, I think it makes more sense for it to be here.

unho replied:
@julen What do you mean with "format spec"? Format for this JSON file?

julen commented on docs/server/commands.rst:
```diff
@@ -346,6 +346,129 @@ By default, :ref:`commands#test_checks` tests all existing checks. When
+The JSON file consists in a list of virtual folder definitions with the
+following fields:
+
+- A mandatory lowercase name for the virtual folder,
```
For reader's parsing ease, you'll want to backtick-quote the attribute names or make a definition list out of them.

unho replied:
Tried backticking, but I ended up using bold. I think it is more readable. Thanks!

unho commented:
Fixed all the nits except the one about where the JSON example is included in the docs.

unho commented:
@julen @dwaynebailey @iafan @jleclanche More reviews and gtms are welcome.

dwaynebailey commented:
The following are results from my testing.  I setup a simple test with a files containing one each: translated, untranslated, fuzzy and obsolete entries.  Placed in directories a, b and c.

- You get a traceback if you don't supply a gaols file, I'd hazard that if it isn't JSON you'd also get a problem.
- Sorting is busted in that we're sorting on priority and state i.e high priority untranslated, then high priority fuzzy, then normal untranslated, normal fuzzy, low, translated low fuzzy.
- When at http://localhost:8000/projects/firefox2/browser/ i.e. project overview there are no priority links to the languages when I've filtered to a goal location.
- Setup three dirs a, b and c. Setup goals priority of c=2.0, b=0.5.  Correct ordering would then be c, a, b.  Yet when stepping through it goes c, b, a

As suggestion to make testing easier: show priority and goal that applies to the unit on the editor.  Then its easy for a tester to evaluate why a unit is being displayed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+
+.. _virtual_folders#attributes:
+
+Virtual folders attributes
```
*folders'*

unho replied:
Changed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+.. _virtual_folders#attributes:
+
+Virtual folders attributes
+-----------------------------------
```
Underline unnecessarily long.

unho replied:
Fixed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+By default virtual folders cannot be browsed: they won't be displayed, but
+they are still used for sorting.
+
+Also the virtual folders can have a description which might be useful to tell
```
*useful to explain the contents of the folder or provide additional instructions*.

unho replied:
Changed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+
+.. _virtual_folders#json-format:
+
+Format for the JSON file
```
If you read through the whole doc, this section sounds like a repetition of *Virtual folders' attributes*. Also, in the first section you miss out part of the information contained in this section. I'd suggest merging both sections.

unho replied:
Fixed.

julen commented on pootle/apps/virtualfolder/management/commands/add_vfolders.py:
```diff
@@ -0,0 +1,128 @@
+                    vfolder.save()
+                    updated_count += 1
+
+        logging.info("\nSucessfully added %d virtual folders.\nSuccessfully "
```
There's a typo (*sucessfully*), but besides, it looks easier to parse an output such as:
```
Added: M
Updated: N
Unchanged: P
```

unho replied:
Changed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+  higher priority,
+- An optional ``true`` or ``false`` indicating if the virtual folder is
+  **browsable** (defaults to ``false``),
+- An optional **description** that can include newlines (that must be escaped),
```
*(must be escaped)*.

unho replied:
Changed.

julen commented on docs/features/virtual_folders.rst:
```diff
@@ -0,0 +1,96 @@
+The JSON file used to import virtual folders consists of a list of virtual
+folder definitions with the following fields:
+
+- A mandatory lowercase **name** for the virtual folder,
```
Why lowercase? What happens when it's not lowercase?

unho replied:
It is converted to lowercase.

Telling people that it must be lowercase reduces the chance of specifying two vfolders with the same lowercased name, thus failing the addition of the second one.

julen replied:
Would it fail? I see name + location are unique together, not name.

unho replied:
@julen It will fail if the location is the same and the names lowercased are the same.

julen replied:
Therefore it won't fail to add *two vfolders with the same lowercased name*.

unho replied:
Only would fail if they also share the same location.

julen commented on pootle/apps/virtualfolder/models.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        if "{LANG}" in self.location and "{PROJ}" in self.location:
+            locations = []
+            temp = ""
```
Why do you need this?

unho replied:
When you save a vfolder, if its location contains `{LANG}` or `{PROJ}` placeholders they have to be replaced by all the existing language and project codes. This is necessary to have all the locations the vfolder applies to, so they can be concatenated with the file paths and thus retrieve the stores (if they exist) so the vfolder can create relationships with their units.

julen replied:
Let me rephrase that: you always destroy the value as soon as you enter the loop, so what's the point of having this value set here?

unho replied:
I got what you meant. It is not necessary. Removing it.

julen commented on pootle/apps/virtualfolder/models.py:
```diff
@@ -0,0 +1,132 @@
+        return ": ".join([self.name, self.location])
+
+    def save(self, *args, **kwargs):
+        if not self.priority > 0:
```
This is model validation, it should go in `clean_fields()` and raise a `ValidationError`.

julen replied:
Since there's no form/`is_valid()` involved, the location is fine, just raise `ValidationError`.

unho replied:
Instead added a `clean_fields()` that raises `ValidationError` and added an explicit call at the beginning of `save()`. I think it helps with keeping `save()` clean.

julen replied:
Both are clean, in any case it'd help if in the future there's a `ModelForm` of any sort :+1:

julen commented on pootle/apps/virtualfolder/models.py:
```diff
@@ -0,0 +1,132 @@
+        retrieve only those with filtering rules that actually match.
+        """
+        return VirtualFolder.objects.filter(
+            units__store__pootle_path__startswith=pootle_path
```
I believe this wouldn't give the desired results when passing a `pootle_path` such as `/projects/<my_project>/<my_directory>/`; would it?

unho replied:
No. This is something that will need to be adjusted when this code is actually used to retrieve the vfolders for such paths. Right now we are only dealing with vfolders within TPs.

julen replied:
So why is it restricted to TPs for now? FWIW going through `/projects/<my_project>/[<my_directory>/]` is a very valid use-case, especially for l10n managers.

unho replied:
@julen It is restricted to TPs for now because TP overview is the only place where we currently use vfolders. When we support other overviews as well, we will have to alter this accordingly.

As yes, that is the translation managers key use case.

julen commented:
Did some initial testing, here are some observations in no particular order:

1. Is there a reason why passing the file via `--filename` is required? `add_vfolders myfile.json` looks natural, especially taking into account a file is the only way to provide input to the command.
2. Providing an invalid JSON file produces:  
  ```
CommandError: Some error occurred while parsing the file
  ```  
  Not very descriptive, plus not sure if this really fails into the `CommandError` category — it's a parsing error after all, and the user wants to know what needs fixing.
3. The `Succesfully parsed '<myfile>.json'` message is superfluous.
4. Some `location` paths seem to be unsupported, yet they are added as `VirtualFolder`s and sorting doesn't work (`/`, `/projects/<project_name>/`).

julen commented:
In the JSON file on the other hand, I believe any fields affecting filtering should go as children of a `filter` key as defined in the spec — it's explicit and easy to understand which filtering rules will be applied, and especially because more filtering options are foreseen.

unho commented:
> Did some initial testing, here are some observations in no particular order:
> 
> 1. Is there a reason why passing the file via `--filename` is required? `add_vfolders myfile.json` looks natural, especially taking into account a file is the only way to provide input to the command.

Fixed.

> 2. Providing an invalid JSON file produces:  
>   ```
> CommandError: Some error occurred while parsing the file
>   ```  
>   Not very descriptive, plus not sure if this really fails into the `CommandError` category — it's a parsing error after all, and the user wants to know what needs fixing.

Altered error handling to raise a `CommandError` with the original message and a hint for failed JSON parsing.

> 3. The `Succesfully parsed '<myfile>.json'` message is superfluous.

Removed.

> 4. Some `location` paths seem to be unsupported, yet they are added as `VirtualFolder`s and sorting doesn't work (`/`, `/projects/<project_name>/`).

Now `/projects/<project_name>/` is transparently converted to `/{LANG}/<project_name>/` when associating the vfolder with the units. If `/` is provided a `ValidationError` is raised asking to provide instead `/{LANG}/{PROJ}/` since it might not be safe to transparently replace the location and also the specified file paths.

unho commented:
> In the JSON file on the other hand, I believe any fields affecting filtering should go as children of a `filter` key as defined in the spec — it's explicit and easy to understand which filtering rules will be applied, and especially because more filtering options are foreseen.

Then the JSON file will need to be altered accordingly when such filtering options are added.

unho commented:
Updated PR addressing most of the issues found. Also removed most of the comments to not overwhelm reviewers.

@julen @dwaynebailey @jleclanche @iafan @ta2-1 Any other issue on this PR?

unho commented:
A fix for the wrong priority ordering on the editor is still when crafted.

unho commented:
@julen @dwaynebailey @jleclanche @iafan @ta2-1 PR ready for final review. Priority sorting on editor now works (despite the ugly hack needed).

julen commented:
> Then the JSON file will need to be altered accordingly when such filtering options are added.

So why not do it right from the beginning and avoid future issues?

julen commented on pootle/apps/virtualfolder/management/commands/add_vfolders.py:
```diff
@@ -0,0 +1,135 @@
+        if not args:
+            raise CommandError("You forgot to provide the mandatory filename.")
+        elif len(args) > 1:
+            logging.warning("You have provided more than one argument. Only "
```
Is this really useful?

unho replied:
No. Will remove it.

julen commented on pootle/core/helpers.py:
```diff
@@ -154,9 +155,17 @@ def get_overview_context(request):
+    filters = {}
+
+    if not isinstance(resource_obj, Store):
+        if VirtualFolder.get_matching_for(request.pootle_path).count():
```
you can combine these into a single `if` expression.

unho replied:
Fixed.

julen commented on pootle/core/browser.py:
```diff
@@ -85,14 +87,14 @@ def get_table_headings(choices):
     """Template variables for each row in the table."""
     return {
         'href': path_obj.get_absolute_url(),
         'href_all': path_obj.get_translate_url(),
```
This should also receive `**kwargs`, I'd say

unho replied:
@julen No. The only kwarg we are passing now is the priority sorting, and we don't want priority sorting when translating all the units.

unho commented:
@julen @dwaynebailey @jleclanche @iafan @ta2-1 Any final review for this PR? Or is it gtm?

dwaynebailey commented:
@unho gtm from my side.

Only one thing though is @julen's comment about filter rules being children which sounds like a sensible suggestion that would allow any other filtering rules to be added without having to change the JSON.  I'm not clear why you're rejecting that idea.

unho commented:
@dwaynebailey I am not rejecting the idea. My point on this is that this should be done once we have a clear idea of how the filtering is going to work. And either way the JSON will be changed.

dwaynebailey commented:
@unho I think you might be missing my point.  Without the children implemented everyone **will** have to change their JSON (as distinct from *the* JSON).  Doing children means that the current file based and whatever new one can live side by side without a change.  Just saving all of us some pain and allowing the two to live together if needed.

That's the reasoning which I'm not seeing being challenged here.

unho commented:
> I think you might be missing my point.  Without the children implemented everyone **will** have to change their JSON (as distinct from *the* JSON).  Doing children means that the current file based and whatever new one can live side by side without a change.  Just saving all of us some pain and allowing the two to live together if needed.

Lets talk with examples. This is the current status:

```json
[
    {
        "name": "current-status",
        "location": "/{LANG}/proj/dir/",
        "files": [
            "fileA.po",
            "subdir/fileB.po"
        ]
    }
]
```

This was @julen proposal:

```json
[
    {
        "name": "julen-proposal",
        "filters": {
            "location": "/{LANG}/proj/dir/",
            "files": [
                "fileA.po",
                "subdir/fileB.po"
            ]
        }
    }
]
```

And this was my proposal for moving the "files" into the "filters" (but keeping "location" as a property of the vfolder):

```json
[
    {
        "name": "my-proposal",
        "location": "/{LANG}/proj/dir/",
        "filters": {
            "files": [
                "fileA.po",
                "subdir/fileB.po"
            ]
        }
    }
]
```

The reason why I am reluctant to move the "files" into the "filters" is that we are not yet sure if when a proper filtering is in place we will have any "files" anywhere. So if it is going to be dropped in the future, why bother now about moving it there?

So until we add more filtering rules, and we are sure we will keep the "files" around (at least in this way), I suggest to keep the current JSON format since it provides the same data with a simpler syntax.

unho commented:
@dwaynebailey @julen Altered the JSON syntax to the following. Does this address your concerns?

```json
[
    {
        "name": "my-proposal",
        "location": "/{LANG}/proj/dir/",
        "filters": {
            "files": [
                "fileA.po",
                "subdir/fileB.po"
            ]
        }
    }
]
```

dwaynebailey commented:
@unho yes that new JSON layout addresses my concerns

unho commented:
@iafan A review from your side would be much appreciated.

unho commented:
@translate/evernote @translate/house People seems to be ok with the new JSON syntax, which was the last blocker. Any other issue or gtm?

iafan commented:
I did a first round of review of this PR on our staging server, which wasn't quite successful. In short, I was unable to successfully add any v-folder description that would show up in the UI. I don't know if this is related to caching or not, but nothing changed from the UI standpoint when I added v-folders via add_vfolders command. I tried simplest scenarios when there's a v-folder inside a folder A, which combines two files (B.po and C.po) located in the same folder A.

Also, the current v-folder functionality is very limited (my assumptions here are based on the documentation and samples alone): one can only provide the set of files to group together. In our case, we can only imagine real use of this feature when one can specify any arbitrary folder (starting with project folders).

It is not obvious how to delete v-folders. When I delete vfolder definitions from the .json file and run add_vfolders, nothing happens. Does add_vfolders (or any other command) support deletion at all?

It is not obvious whether one needs to include leading and/or trailing slash in `location` path. The command gives no errors.

Side notes: I noticed these regressions:
1) the sidebar is always visible (can be expanded/collapsed), even it doesn't have any content in it. In the master branch code, it is not available if it has nothing to display.
2) the sidebar height depends on the size of the content (the number of rows in the browsing table). It should take the full height of the content area.

To sum it up, this definitely needs another round of testing and bugfixing before the feature goes live.

julen commented:
> Side notes: I noticed these regressions:
> 1) the sidebar is always visible (can be expanded/collapsed), even it doesn't have any content in it. In the master branch code, it is not available if it has nothing to display.
> 2) the sidebar height depends on the size of the content (the number of rows in the browsing table). It should take the full height of the content area.

These are very likely remnants of a previous state of the PR branch so it's safe to ignore it. I recreated assets just in case and cannot see this happening on our staging server.

iafan commented:
Thanks, @julen. Are you able to setup v-folders on your side? I think I tried to closely follow the sample JSON provided, but have no luck seeing the v-folders in the browser

iafan commented:
@julen,

I can actaully see the empty but open side panel if I go to https://translate.stage.evernote.com/ru/website3_evernote/home/ (again, this might be some side effect of caching, but worth checking on your side)

dwaynebailey commented:
@iafan this first stage commit has no UI.  The change forces editing based on v-folder priority.  The second stage, not yet PR'd, adds the UI for the TP level.

iafan commented:
@dwaynebailey, so how one is supposed to test this? It makes sense to add UI to this PR as well, and test the feature as a whole, IMO.

julen commented:
> @julen,
> 
> I can actaully see the empty but open side panel if I go to https://translate.stage.evernote.com/ru/website3_evernote/home/ (again, this might be some side effect of caching, but worth checking on your side)

@iafan I realize some setting changes to disable the imports/exports were uncommitted on myside, and that's fixed now. In any case this displays that imports/exports in master is not in good shape, and I already reported this on IRC.

unho commented on pootle/apps/virtualfolder/models.py:
```diff
@@ -0,0 +1,146 @@
+            return locations
+        elif "{LANG}" in self.location:
+            return [self.location.replace("{LANG}", lang.code)
+                    for lang in Language.objects.all()]
```
This should use only current project languages.

unho commented on pootle/apps/virtualfolder/models.py:
```diff
@@ -0,0 +1,146 @@
+                    for lang in Language.objects.all()]
+        elif "{PROJ}" in self.location:
+            return [self.location.replace("{PROJ}", proj.code)
+                    for proj in Project.objects.all()]
```
This should use only current language projects.

unho commented:
It seems that I missed the party.

> I did a first round of review of this PR on our staging server, which wasn't quite successful. In short, I was unable to successfully add any v-folder description that would show up in the UI. I don't know if this is related to caching or not, but nothing changed from the UI standpoint when I added v-folders via add_vfolders command. I tried simplest scenarios when there's a v-folder inside a folder A, which combines two files (B.po and C.po) located in the same folder A.

No descriptions are being displayed as it is not yet clear to me where does should be displayed.

> Also, the current v-folder functionality is very limited (my assumptions here are based on the documentation and samples alone): one can only provide the set of files to group together. In our case, we can only imagine real use of this feature when one can specify any arbitrary folder (starting with project folders).

This is just the first step. In the future you can expand this as necessary to support more filtering methods. For now using stores just recovers our lost goals functionality.

> It is not obvious how to delete v-folders. When I delete vfolder definitions from the .json file and run add_vfolders, nothing happens. Does add_vfolders (or any other command) support deletion at all?

No. This is not yet implemented.

You can instead hide vfolders by setting them to not being browsable. Changing vfolders is supported through the `add_vfolders` command by just providing a JSON file altered accordingly.

> It is not obvious whether one needs to include leading and/or trailing slash in `location` path. The command gives no errors.

I was assuming that people would check the docs and the sample JSON in case of doubt. As you can see the sample JSON always includes the trailing slashes. Just checked the code and it never complains, but in some cases the absence of trailing slash might result on not getting the desired results on filtering.

> Side notes: I noticed these regressions:
> 1) the sidebar is always visible (can be expanded/collapsed), even it doesn't have any content in it. In the master branch code, it is not available if it has nothing to display.
> 2) the sidebar height depends on the size of the content (the number of rows in the browsing table). It should take the full height of the content area.

No change is done in this PR to the sidebar, so these should be issues already present on master.

> To sum it up, this definitely needs another round of testing and bugfixing before the feature goes live.

I don't think this PR has bugs. Re testing, see below.

> @dwaynebailey, so how one is supposed to test this? It makes sense to add UI to this PR as well, and test the feature as a whole, IMO.

Please see the testing steps provided on comment https://github.com/translate/pootle/pull/3600#issuecomment-76679565 I am aware that lack of UI difficults too much the testing, but the steps in that comment should be enough. If they are not enough we can talk about appending here some of the commits already finished that add UI bits, despite that I would prefer to land the current set of commits since has been tested and reworked several times instead of adding more commits that might need to be reworked more times thus delaying this PR even more.

unho commented:
@iafan I agree we need a way to remove a vfolder.

IMHO considering that a vfolder that is not on the JSON means that it should be removed is a bad idea. I mean that if you have 100 vfolders you usually going to use the JSON to just alter some vfolders (lets say 5 of them) and having to deal with a JSON with 100 vfolder is not comfortable. Besides that having a `add_vfolders` that allows to remove them is kind of confusing.

Another option is to add a new management command `del_vfolder` which accepts some data identifying a vfolder and removes it. I think it is better for it to remove only a single vfolder at a time. Since vfolders have no unique identifiers (languages and projects have their codes) we will have to provide a name and a location, so we might end with something like:

```shell
$ ./manage.py del_vfolder my-vfolder /{LANG}/proj/dir/subdir/subsubdir/
```

which I am not sure if can be considered simple and easy.

Anyway whichever way we take re deleting, IMHO it is less priority than providing a working UI for translators to work with vfolders. Despite it is an ugly workaround, you can always hide vfolders by setting them to not being browsable.

dwaynebailey commented:
@unho can you create an issue to track the del_vfolder command.  I think we can delay it with the 1.0 and not browesable hack.  But we do need to fix it to complete the feature.

dwaynebailey commented:
@iafan the way I tested was as following.  I created 3 simple 3 unit PO files (translated, untranslated, fuzzy) a.po, b.po, c.po.  I created vfolder that set 2.0 for b.po and 0.5 priorities for a.po.  Used the editor to ensure that the files appear in the order specified which would be b.po, c.po, a.po.

This first implementation is designed simply to ensure the backend for storing goals is good and that priority sorting of gaols works correctly.

unho commented:
@dwaynebailey Done: https://github.com/translate/pootle/issues/3634

iafan commented:
> This first implementation is designed simply to ensure the backend for storing goals is good and that priority sorting of gaols works correctly.

Then I can delegate the review to @julen. If this feature is not visible and has not reached the level it can be of real use to us (Evernote), I'm ok to have it merged if Julen confirms that it looks sane from the backend side. I'll be happy to review this once the UI part has landed, and once we can start testing this on some real-life scenarios (for us the first most obvious case will be to group projects together, which needs the filters to support paths, not individual .po files).

iafan commented:
> MHO considering that a vfolder that is not on the JSON means that it should be removed is a bad idea. I mean that if you have 100 vfolders you usually going to use the JSON to just alter some vfolders (lets say 5 of them) and having to deal with a JSON with 100 vfolder is not comfortable. Besides that having a add_vfolders that allows to remove them is kind of confusing.

@unho, I think I disagree on this. From my perspective, it would be much easier to have one JSON file which fully represents the config for all v-folders, and maintain it, adding or deleting entries from it. Then a single command, say, `update_vfolders` would synchronize this JSON with the database, creating, updating or deleting v-folders accordingly.

Otherwise, you're going to have two commands (`add_vfolders` and `del_vfolders`) which you have to remember about and maintain data for, and deletion of vfolders will be a much less intuitive task than just deleting relevant entries from a single JSON.

unho commented:
>> IMHO considering that a vfolder that is not on the JSON means that it should be removed is a bad idea. I mean that if you have 100 vfolders you usually going to use the JSON to just alter some vfolders (lets say 5 of them) and having to deal with a JSON with 100 vfolder is not comfortable. Besides that having a add_vfolders that allows to remove them is kind of confusing.

> @unho, I think I disagree on this. From my perspective, it would be much easier to have one JSON file which fully represents the config for all v-folders, and maintain it, adding or deleting entries from it. Then a single command, say, `update_vfolders` would synchronize this JSON with the database, creating, updating or deleting v-folders accordingly.

> Otherwise, you're going to have two commands (`add_vfolders` and `del_vfolders`) which you have to remember about and maintain data for, and deletion of vfolders will be a much less intuitive task than just deleting relevant entries from a single JSON.

I see your point, but I was thinking on a several admins scenario. Maybe yours is better.

@dwaynebailey any input on this? 

> (for us the first most obvious case will be to group projects together, which needs the filters to support paths, not individual .po files).

@iafan Can you explain this a bit more?

iafan commented:
> I see your point, but I was thinking on a several admins scenario.

I think this is pretty much similar to regular Pootle configs. We have several Pootle admins, and they have an access to the Pootle configuration file, which we keep in Git. The same thing is applicable to this JSON config — especially if there are multiple people working with it — it makes sensу to keep it in version control (and this is where its history will be stored), and then apply it as a whole (e.g. as a part of deployment script) to the server.

iafan commented:
> (for us the first most obvious case will be to group projects together, which needs the filters to support paths, not individual .po files).
> @iafan Can you explain this a bit more?

This is pretty straightforward: imagine you can set up any path prefix in your `filters > files` section of the v-folder JSON entry. This way you would include all files down that tree, and this should work right from any point in the tree. Consider this example:
```
    {
        "name": "ios_skitch_evernote_bundle",
        "location": "/{LANG}/",
        "priority": 1,
        "is_browsable": true,
        "description": "Skitch (iOS) — all related projects",
        "filters": {
            "files": [
                "ios_skitch_evernote",
                "common_skitch_evernote"
            ]
        }
    },
```

This is a v-folder that lives inside each language folder, and which groups two translation projects together, identified by their relative paths: `ios_skitch_evernote` and `common_skitch_evernote`.

unho commented:
@iafan Implementing filtering by path besides filtering by specific filenames looks easy to implement. Added issue #3643 to track it.

unho commented:
@iafan Lets move discussion re deleting vfolders to issue #3634.
